### ISSUE TITLE
nosetests fails to discover tests with nosedep enabled

### ISSUE THREAD

**Issue by OscarL** [Zitrax/nose-dep#3]
I'm using nose 1.3.7 to run several tests, and I've just installed nose-dep 0.2 (nose and nosedep installed in a virtualenv).

I have not changed my tests yet to use `@depends`.

My directory layout is basically like this:

```
package/
    __init__.py
    some_code.py
    tests/
        foo/
            __init__.py
            test_package.py
```

If I do:

```
~/package$ export PYTHONPATH=~/package
~/package$ nosetests tests
....
Ran 10 tests in 10.000s
```

All my tests are run properly. But as soon as I add --with-nosedep:

```
~/package$ nosetests --with-nosedep tests
....
Ran 0 tests in 0.000s
```

The following also works

```
~/package$ nosetests tests/foo/test_package.py
```

But this:

```
~/package$ nosetests --with-nosedep tests/foo/test_package.py
```

I get:

```
Traceback (most recent call last):
  File "/home/user/package/venv/bin/nosetests", line 11, in <module>
    sys.exit(run_exit())
  File "/home/user/package/venv/local/lib/python2.7/site-packages/nose/core.py", line 121, in __init__
    **extra_args)
  File "/usr/lib/python2.7/unittest/main.py", line 95, in __init__
    self.runTests()
  File "/home/user/package/venv/local/lib/python2.7/site-packages/nose/core.py", line 207, in runTests
    result = self.testRunner.run(self.test)
  File "/home/user/package/venv/local/lib/python2.7/site-packages/nose/core.py", line 50, in run
    wrapper = self.config.plugins.prepareTest(test)
  File "/home/user/package/venv/local/lib/python2.7/site-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/home/user/package/venv/local/lib/python2.7/site-packages/nose/plugins/manager.py", line 167, in simple
    result = meth(*arg, **kw)
  File "/home/user/package/venv/local/lib/python2.7/site-packages/nosedep.py", line 235, in prepareTest
    reraise(t, v, tb)
  File "/home/user/package/venv/local/lib/python2.7/site-packages/nosedep.py", line 230, in prepareTest
    all_tests[tt.test.test.__name__] = tt
AttributeError: 'Failure' object has no attribute 'test' - Due to: Failure: OSError (No such file /home/user/package/tests/foo/test_package)
```

Besides the somehow convoluted project layout (can't help it, not my project), am I missing something or doing it wrong? Or is this an issue on nosedep?

Thanks for sharing your work, looking forward to be able to use it! :-D

**Comment by Zitrax** [comment]
Hi Oscar, thanks for trying it out and reporting.

nosedep has not yet been tried on many projects so there might be some difference that cause problems. But it would be very good to figure out what is going wrong here.

I see one difference here compared to how my project is run, you are using `nosetests --with-nosedep file.py` while I have been running `file.py --with-nosedep` where my test file contains a main section which launch nosetests internally. It should definitely work the way you are launching too so I'll need to investigate why that is not the case. ( A quick local check seemed to reproduce your issue )

I'll investigate this more in the evening or in the weekend. Is the project you use available for easier debugging/verification ?

**Comment by OscarL** [comment]
Hello Zitrax.

Thank you for taking time to investigate this.

Sadly, the project I'm using is private. But besides the code layout, the only thing worth mentioning about the tests code is:

It only uses unittest.TestCase for derived test cases classes (no TestSuites, no custom runners, not even unittest.main()). It relies solely on nose's for test discovery. It mixes plain asserts with self.assertXXX() from unittest.

Basically, you could just use the following:

``` Python
import unittest


class SomeTest(unittest.TestCase):
    def setUp(self):
        print 'setUp SomeTest'

    def tearDown(self):
        print 'tearDown SomeTest'

    def test_is_computer_on(self):
        self.assertIsNotNone(1)

    def test_is_computer_on_fire(self):
        assert 1 == 1
```

Thanks again for your work and your time!

**Comment by Zitrax** [comment]
Thanks for the info, I'll notify you again when I have looked into it.

**Comment by Zitrax** [comment]
I can easily reproduce some issues, the above commit was just a partial fix. Further changes are needed...

**Comment by Zitrax** [comment]
I did some further changes. The two commits were specifically targeted at:
- Tests inheriting from `unittest.TestCase`.
- When specifying a file on the command line, like `test.py`.
- When spcifying a class on the command line like `test.py:Test`.

There are a big number of ways to run tests so I have definitely not covered all possible ways. But added tests for the methods above.

Also travis runs have one test failure while they works fine for me locally, might be os/platform related - so will have to  look into that before considering a new release.

But would be nice if you could see if the changes helped, if you want to try trunk you can clone the repo and then run `python setup.py install` to install the source version as a package.

**Comment by OscarL** [comment]
I did a quick inspection. In both the following ways...

```
$ nosetests --with-nosedep tests
$ nosetests --with-nosedep tests/foo/test_package.py
```

It works!

As in: it runs the same tests it does if I don't specify "--with-nosedep", which is great!

I can now start trying to use `@depends` on the tests.

Thanks a lot!

PS: I'll report back if I find anything else.

**Comment by OscarL** [comment]
Ok. I've added some `@depends(after='test_foobar')` to a couple of methods in a simple class derived from unittest.TestCase.

I force "test_foobar" to fail (assert False, or raise Exception), but the rest of the tests keep getting executed (and passing in my case).

Should I open a new issue for that?

Also, is something like the following even possible?:

``` Python
@depends(after='SomeOtherTestCase')
class SomeTests(unittest.TestCase):
    ....
```

EDIT: This is extrange, but might help you perhaps.

I've decorated one of the methods with: `@depends(after='not_existing_test')`, and all the tests ran just the same. 

But if I put this in the same file (as top-level code):

``` Python
@depends(after='not_existing_test')
def test_dfds_d():
    pass
```

That test fails with `Exception: Required test 'not_existing_test' did not run (does it exist?)` as expected.

**Comment by Zitrax** [comment]
I have not targeted or tested to decorate a class so if that had worked it would just have been luck :) Possibly it could be added later - not sure how easy it would be without investigating further. Until then I guess it would have been nice to raise an exception that it's not supported.

Regarding the problem running the test even though the dependency failed; yes I can reproduce it too when the class inherits from `unittest.TestCase`. If I remove that inheritance it works as expected. I don't know if you need that inheritance (nose tests does not require it). Still it would be good if it worked of course, I'll have a look.

Thanks again for the feedback.

**Comment by Zitrax** [comment]
Added patch for the `unittest.TestCase` dependency issue. Could you try it ?

**Comment by OscarL** [comment]
Alright, just ended testing your latest changes. All the mentioned problems seem to be gone! Well done!!!

I can now use `@depends` on test methods in instances that inherit from unittest.TestCase! (that's how this project does it, it's using some of the utility self.assertXXX() methods in many places).

So, great work Zitrax! Thanks!

FYI, the other "thing": decorating a whole class like this:

```
@depends(after='SomeOtherTestCase')
class SomeTests(unittest.TestCase):
    ....
```

Simply makes, silently, none of SomeTest's tests to be run, so I guess that it would be better to raise an exception in that case (as you mentioned) to avoid people like me getting confused :-D

**Comment by Zitrax** [comment]
Great, glad it worked :)

I'll have a look at raising for class decorations too. But I'll close this issue now since the main problems are fixed.

**Comment by Zitrax** [comment]
Added the class decorator check in 27cca51025a2cce4f935b55c04ea6a9e9d92112a.
